### PR TITLE
Treat an ERR_MEM on send as non-fatal on LwIP.

### LINKED ISSUE #21671: Crit-err 3000001 on retransmiting CASE Sigma2
#### Problem

In CASE process (during device commissioning), the comissionee should reply Sigma2 to comissioner.
But there is a little chance that the comissionee failed sent Sigma2 message.

It happens with a very low chance (<1%), but we have catched the problem **twice**. Both of them have exactly same phenomenon:

1. Sent Sigma2 msg first time, no error reported, but the comissioner does not received
2. Retransmit Sigma2 after 360ms, no error reported, but the comissioner still does not received
3. Retransmit Sigma2 third times, an error `Crit-err 3000001` reported, retransmition stopped, and the comissioner still does not received

After that, device commissioning failed due to timeout in the comissioner side.

The error code `3000001` means `ERR_MEM`, but some log messages showed that free heap size is enough to handle network message.

 * system configuration
  Comissioner: Android (commit: e556daac2e1ed3a141034a6dcc7e410e4cd1f8f6)
  Comissionee: ESP32 (commit: e556daac2e1ed3a141034a6dcc7e410e4cd1f8f6)

* Logs
Comissioner: [log2.txt](https://github.com/project-chip/connectedhomeip/files/9266844/log2.txt)
Comissionee: [Device_2022-08-05_15_58_32.log](https://github.com/project-chip/connectedhomeip/files/9266841/Device_2022-08-05_15_58_32.log)

```
I (25176) chip[IN]: Prepared unauthenticated message 0x3ffc6af4 to 0x0000000000000000 (0)  of type 0x31 and protocolId (0, 0) on exchange 5359r with MessageCounter:112355573.
I (25196) chip[IN]: Sending unauthenticated msg 0x3ffc6af4 with MessageCounter:112355573 to 0x0000000000000000 at monotonic time: 0000000000005F7E msec
I (25206) chip[SC]: Sent Sigma2 msg
E (25206) chip[DL]: Long dispatch time: 1400 ms, for event type 3
I (25216) chip[DL]: IP_EVENT_GOT_IP6
I (25226) chip[DL]: IPv6 addr available. Ready on WIFI_STA_DEF interface: fe80:0000:0000:0000:260a:c4ff:fe49:41b4
I (25236) app-devicecallbacks: Current free heap: 96384

I (25236) chip[DIS]: Updating services using commissioning mode 0
I (25246) chip[DIS]: CHIP minimal mDNS started advertising.
I (25256) chip[DIS]: Advertise operational node 2DD462755BFF7F02-061ABCD5F72C7C00
I (25266) chip[DIS]: CHIP minimal mDNS configured as 'Operational device'.
I (25286) chip[DIS]: mDNS service published: _matter._tcp
I (25286) app-devicecallbacks: Current free heap: 94084

I (25386) CHIP[DL]: Write request received for CHIPoBLE RX characteristic con 0 12
I (25556) chip[IN]: Sending unauthenticated msg 0x3ffc6af4 with MessageCounter:112355573 to 0x0000000000000000 at monotonic time: 00000000000060E6 msec
I (25896) chip[IN]: Sending unauthenticated msg 0x3ffc6af4 with MessageCounter:112355573 to 0x0000000000000000 at monotonic time: 000000000000623F msec
E (25916) chip[EM]: Crit-err 3000001 when sending CHIP MessageCounter:112355573 on exchange 5359r, send tries: 2
```

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
That can be caused by a transient lack of TX buffers.  We should just
try again via MRP.

Fixes https://github.com/project-chip/connectedhomeip/issues/21671

#### Problem
Errors that should be transient end up being fatal.

#### Change overview
Treat ERR_MEM on send as a transient error when using LwIP.

#### Testing
Needs real-world testing.  I asked for some in https://github.com/project-chip/connectedhomeip/issues/21671 but no responses so far.

### CODE REVIEW

File: src/messaging/ReliableMessageMgr.cpp [hunk 1]
```diff
@@ -416,6 +417,32 @@ void ReliableMessageMgr::RegisterSessionUpdateDelegate(SessionUpdateDelegate * s
     mSessionUpdateDelegate = sessionUpdateDelegate;
 }
 
+CHIP_ERROR ReliableMessageMgr::MapSendError(CHIP_ERROR error, uint16_t exchangeId, bool isInitiator)
+{
+    if (error == CHIP_ERROR_POSIX(ENOBUFS)
```
Comment (tcarmelveilleux) [inline src/messaging/ReliableMessageMgr.cpp:422]: This should not be looked-at for LWIP. I think we need to have less platform-specifics in here.

Please use #if and #else here

Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageMgr.cpp:422]: I can do that.  I was a little surprised we even compiled CHIP_ERROR_POSIX bits on LWIP.....

Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageMgr.cpp]: Done.

**Review by tcarmelveilleux — APPROVED** [review]

**Comment by andy31415** [comment]
Accepted for 1.0 as high value fix for LWIP

**Review by jmartinez-silabs — APPROVED** [review]

**Pull request merged**